### PR TITLE
[dsl.pipeline] Pipeline support variable args as pipeline input

### PR DESCRIPTION
**Pull request by lalala123123**
# Description

Pipeline support variable args as pipeline inputs. Custorm can use *args and *kwargs to set the variable inputs. The value in *args will be named as args_[index]

```python
@dsl.pipeline
def pipeline_func(component_in_number: int, component_in_path: Input, **kwargs):
    node = component_func1(component_in_number=component_in_number, component_in_path=component_in_path)
    node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
                                  component_in_path=kwargs["component_in_path1"])

pipeline = pipeline_func(10, data, component_in_number1=11, component_in_path1=data)
```

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py [hunk 1]
```diff
@@ -2086,3 +2086,51 @@ def pipeline_func(job_in_number, job_in_path, component_func):
...
+            node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
+            node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
+                                          component_in_path=kwargs["component_in_path1"])
+            node_with_arg_kwarg = pipeline_with_variable_args(*args, **kwargs)
+
+        pipeline = root_pipeline(10, data, 11, data, component_in_number1=11, component_in_path1=data)
+
+        assert len(pipeline.inputs) == 6
+        arg_names = ['component_in_number', 'component_in_path', 'component_in_number1',
+                     'component_in_path1', 'args_0', 'args_1']
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2113]: Add some case that there is conflict parameter name
def pipepline(args_0, *args)

def pipepline(*args, **kwargs):
   pass
pipeline(list, arg_0 =xx)

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py [hunk 2]
```diff
@@ -2086,3 +2086,51 @@ def pipeline_func(job_in_number, job_in_path, component_func):
...
+        node_with_arg_kwarg = pipeline.jobs['node_with_arg_kwarg']
+        assert len(node_with_arg_kwarg.inputs) == 4
+        arg_names = ['component_in_number1', 'component_in_path1', 'args_0', 'args_1']
+        assert all([input._data._name in arg_names for input in node_with_arg_kwarg.inputs.values()])
+
+        with mock.patch(
+            "azure.ai.ml.entities._job.pipeline.pipeline_job.is_private_preview_enabled",
+            return_value=True
+        ):
+            with pytest.raises(UnsupportedParameterKindError,
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2134]: what's the behavior if we do mldesigenr compile

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py [hunk 3]
```diff
@@ -227,33 +235,39 @@ def _validate_args(func, args, kwargs, non_pipeline_inputs):
...
+
+        min_num = len(empty_parameters)
+        max_num = len(named_parameters)
+        if len(args) > max_num:
+            raise TooManyPositionalArgsError(func.__name__, min_num, max_num, len(args))
+
+    func_args, provided_args = list(args), OrderedDict({})
+    provided_kwargs = OrderedDict({param.name: func_args.pop(0) for param in named_parameters if len(func_args) > 0})
+    for index, arg in enumerate(func_args):
+        provided_args[f"args_{index}"] = func_args[index]
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py:261]: do we need to hard code args_?

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py [hunk 4]
```diff
@@ -2086,3 +2086,51 @@ def pipeline_func(job_in_number, job_in_path, component_func):
...
+            node = component_func1(component_in_number=component_in_number, component_in_path=component_in_path)
+            node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
+            node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
+                                          component_in_path=kwargs["component_in_path1"])
+            node_with_arg_kwarg = pipeline_with_variable_args(*args, **kwargs)
+
+        pipeline = root_pipeline(10, data, 11, data, component_in_number1=11, component_in_path1=data)
+
+        assert len(pipeline.inputs) == 6
+        arg_names = ['component_in_number', 'component_in_path', 'component_in_number1',
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2112]: may put full expected inputs yamls here. We would like to test on extra metas like: type, default, description, etc..

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py [hunk 5]
```diff
@@ -2086,3 +2086,51 @@ def pipeline_func(job_in_number, job_in_path, component_func):
...
         assert len(pipeline.jobs) == 2
         assert component_func2.name in pipeline.jobs
+
+    def test_pipeline_with_variable_inputs(self):
+        path = "./tests/test_configs/components/helloworld_component.yml"
+        component_func1 = load_component(path)
+        data = Data(name="test", version="1", type=AssetTypes.MLTABLE)
+
+        @dsl.pipeline
+        def pipeline_with_variable_args(*args, **kwargs):
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2096]: add case that *args -> *cus_args

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py (hunk 3 above)
Comment (brynn-code) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py:261]: you could refer to 1.5 _validate_args, just copy the whole `_validate_args`, it should support variable parameter, I think we don't need args_index as key, just get the param from function.

Comment (lalala123123) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py:261]: Yes, it's possible to support variable parameters by removing some validations refer to v1.5 _validate_args.
But these variable parameters are also pipeline inputs and still need a key as the input name

Comment (lalala123123) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py:261]: I will change the hard code args_ to {variable_args_name}_{index}

Comment (lalala123123) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py]: change the parameter name to {variable_args_name}_{index}

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py (hunk 1 above)
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py]: Add the test case that duplicate parameter name

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py (hunk 5 above)
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2096]: Add test case about other variable args name

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py (hunk 4 above)
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py]: assert full expected inputs yaml

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py [hunk 6]
```diff
@@ -164,13 +165,20 @@ def wrapper(*args, **kwargs) -> PipelineJob:
             # Because we only want to enable dsl settings on top level pipeline
             _dsl_settings_stack.push()  # use this stack to track on_init/on_finalize settings
             try:
-                provided_positional_args = _validate_args(func, args, kwargs, non_pipeline_inputs)
                 # Convert args to kwargs
-                kwargs.update(provided_positional_args)
+                provided_positional_args, provided_positional_kwargs = _validate_args(func, args, kwargs,
```
[the remark this answers is not in the document]
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py:168]: We need to keep the args list for the case that customer uses args index to get arg value.
```python
@dsl.pipeline
def pipeline_with_variable_args(*args, **kwargs):
      node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
```

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_component_builder.py [hunk 7]
```diff
@@ -425,4 +449,6 @@ def _wrap_pipeline_parameter(key, value, group_names=None):
         group_names.append(key)
         return _GroupAttrDict({k: _wrap_pipeline_parameter(k, v, group_names=group_names) for k, v in value.items()})
     # Note: here we build PipelineInput to mark this input as a data binding.
+    if isinstance(value, PipelineInput):
```
[the remark this answers is not in the document]
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_component_builder.py]: Remove it

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py [hunk 8]
```diff
@@ -2086,3 +2086,51 @@ def pipeline_func(job_in_number, job_in_path, component_func):
...
+            node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
+            node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
+                                          component_in_path=kwargs["component_in_path1"])
+
+        @dsl.pipeline
+        def root_pipeline(component_in_number: int, component_in_path: Input, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2108]: Add test case to extract docstring for args and kwargs.

File: sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py (hunk 2 above)
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2290]: If the user indexes by index or key in the pipeline code, it will raise error when execute pipeline code.
```
Traceback (most recent call last):
  File "d:\project\azure-sdk-for-python\sdk\ml\azure-ai-ml\azure\ai\ml\dsl\_pipeline_component_builder.py", line 215, in build
    outputs = self.func(*pipeline_args, **pipeline_kwargs)
  File "D:/Project/azure-sdk-for-python/sdk/ml/azure-ai-ml/tests/test.py", line 50, in root_pipeline
    node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
IndexError: tuple index out of range
```
```
Traceback (most recent call last):
  File "d:\project\azure-sdk-for-python\sdk\ml\azure-ai-ml\azure\ai\ml\dsl\_pipeline_component_builder.py", line 215, in build
    outputs = self.func(*pipeline_args, **pipeline_kwargs)
  File "D:/Project/azure-sdk-for-python/sdk/ml/azure-ai-ml/tests/test.py", line 51, in root_pipeline
    node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
KeyError: 'component_in_number1'
```

Comment (wangchao1230) [reply sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py:2290]: could we raise error before execute user's code?

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py:167]: @lalala123123 then what will the input name be? we have a default value for it?

**Comment by wangchao1230** [comment]
please take a look at the conflicts

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lalala123123) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_pipeline_decorator.py]: Remove support for *args.

**Review by wangchao1230 — APPROVED** [review]
